ext/odbc/php_odbc.c

```diff
@@ -95,19 +101,29 @@ static zend_function *odbc_connection_get_constructor(zend_object *object) {
 
 static void odbc_link_free(odbc_link *link)
 {
-	/* If aborted via timer expiration, don't try to call any unixODBC function */
-	if (!(PG(connection_status) & PHP_CONNECTION_TIMEOUT)) {
-		safe_odbc_disconnect(link->connection->hdbc);
-		SQLFreeConnect(link->connection->hdbc);
-		SQLFreeEnv(link->connection->henv);
-	}
-	efree(link->connection);
-	link->connection = NULL;
-	ODBCG(num_links)--;
-	if (!link->persistent) {
+	ZEND_ASSERT(link->connection && "link has already been closed");
+
+	if (link->persistent) {
+		zend_hash_del(&EG(persistent_list), link->hash);
+	} else {
```

Comment:
`_close_odbc_pconn()` will free most of the stuff when the resource is freed, so we have to avoid doing so  again here, that's why I moved these pieces of code in the `else` block.